cloudinit/config/cc_ssh.py

```diff
@@ -258,9 +260,26 @@ def handle(name: str, cfg: Config, cloud: Cloud, args: list) -> None:
         genkeys = util.get_cfg_option_list(
             cfg, "ssh_genkeytypes", GENERATE_KEY_NAMES
         )
+        # remove keys that are not supported in fips mode if its enabled
+        key_names = (
+            genkeys
+            if not util.fips_enabled()
+            else [
+                names
+                for names in genkeys
+                if names not in FIPS_UNSUPPORTED_KEY_NAMES
+            ]
+        )
+        skipped_keys = set(genkeys).difference(key_names)
+        if skipped_keys:
+            LOG.debug(
+                "skipping keys that are not supported in fips mode: %s",
+                ",".join(skipped_keys),
+            )
```

Comment:
On Ubuntu FIPS (18.04) 4.15 based kernel I'm seeing that ssh-keygen of rsa, ecdsa and ed25519 succeed, and only dsa type fails(See log below).

 I'm wondering if, instead of avoiding the attempt to generate certain keys, we instead want to conditionally process our exception handling `if (fips_enabled and keytype in FIPS_UNSUPPORTED_KEY_NAMES)` avoid calling `log.exec(... Failed generating key type`  but instead just `LOG.debug('Unable to generate key type {keytype} because FIPS is enabled')`


```
Ubuntu FIPS 4.15 based on Ubuntu 18.04 
$ egrep 'WARN|ssh-keygen|Stderr:' /var/log/cloud-init.log
Stderr: mount: /run/cloud-init/tmp/tmpenfrf09s: no medium found on /dev/sr0.
2023-04-28 03:53:29,950 - subp.py[DEBUG]: Running command ['ssh-keygen', '-t', 'rsa', '-N', '', '-f', '/etc/ssh/ssh_host_rsa_key'] with allowed return codes [0] (shell=False, capture=True)
2023-04-28 03:53:30,114 - subp.py[DEBUG]: Running command ['ssh-keygen', '-t', 'dsa', '-N', '', '-f', '/etc/ssh/ssh_host_dsa_key'] with allowed return codes [0] (shell=False, capture=True)
2023-04-28 03:53:30,141 - util.py[WARNING]: Failed generating key type dsa to file /etc/ssh/ssh_host_dsa_key
Command: ['ssh-keygen', '-t', 'dsa', '-N', '', '-f', '/etc/ssh/ssh_host_dsa_key']
Stderr: DSA keys are not allowed in FIPS mode
2023-04-28 03:53:30,148 - subp.py[DEBUG]: Running command ['ssh-keygen', '-t', 'ecdsa', '-N', '', '-f', '/etc/ssh/ssh_host_ecdsa_key'] with allowed return codes [0] (shell=False, capture=True)
2023-04-28 03:53:30,177 - subp.py[DEBUG]: Running command ['ssh-keygen', '-t', 'ed25519', '-N', '', '-f', '/etc/ssh/ssh_host_ed25519_key'] with allowed return codes [0] (shell=False, capture=True)
```

Reply:
As per our doc here: https://access.redhat.com/solutions/3643252 both dsa and ed25519 keys are not allowed in fips mode. I will check if there is a more standard spec on this.

Reply:
https://csrc.nist.gov/CSRC/media/projects/cryptographic-module-validation-program/documents/security-policies/140sp2472.pdf I found this in section 9.1 under "In addition to the configuration of the kernel, the OpenSSH client configuration should contain the following rules:"

Remove or do not configure Ed25519 host keys as Ed25519 keys are not allowed in FIPS Approved
mode. The crypto office should inform the user to not use Ed25519 keys for user authentication.

Reply:
In the bug report I have added the specific commit in fedora that adds the error message in ssh-keygen. So all recent redhat distros are affected. Ubuntu may not be affected by the same way.

Reply:
@ani-sinha I got confirmation that minimally we don't want cloud-init broadly to generate keys unless permitted due to FIPS either being disabled or not present as generating those keys results in bringing that system into non-compliance with FIPS anyway (which you already explained). Aside from cloud-init behavior, we'll take homework to look at what else may need to be in place to ensure a more noisy indication on Ubuntu if entering into non-FIPS compliant mode due to key generation. Ideally I'd like to see something from FIPS solutions that would advertize what certification/policy of FIPS a platform adheres to if there is possibility that newer certifications have change in behavior for cloud-init in terms of key type generation.